### DISCUSSION TITLE
Odd clingo behavior --- brew disagrees with itself over what version is installed

### DISCUSSION THREAD

**Discussion by jasonhemann** [Homebrew/discussions#216] [Casks] [answered]
Moved from https://github.com/Homebrew/homebrew-cask/issues/94074

General troubleshooting steps
[X ] I understand that if I ignore these instructions, my issue may be closed without review.
[X ] I have retried my command with --force.
[X ] I ran brew update-reset && brew update and retried my command.
[X ] I ran brew doctor, fixed as many issues as possible and retried my command.
[X ] I have checked the instructions for reporting bugs.
[X ] I made doubly sure this is not a checksum does not match error.
Description of issue
I tried to upgrade clingo. Brew tells me that it's already installed, and to upgrade clingo. This seems kafkaesque.

Stranger still, now that I've gone through as much of the self-diagnosis steps as I can, I've seen the following. Notice the version numbers actually change.

08:19:08$ brew install clingo
==> Downloading https://homebrew.bintray.com/bottles/clingo-5.4.0_2.catalina.bottle.tar.gz
Already downloaded: /Users/me/Library/Caches/Homebrew/downloads/84f6fc12ad562a91cee3984c937f708316a1e4e3b1634f7f95fe6f71f817f238--clingo-5.4.0_2.catalina.bottle.tar.gz
Error: clingo 3.3.2 is already installed
To upgrade to 5.4.0_2, run `brew upgrade clingo`.
08:19:30$ brew upgrade clingo
==> Upgrading 1 outdated package:
clingo 5.4.0_1 -> 5.4.0_2
==> Upgrading clingo 5.4.0_1 -> 5.4.0_2
==> Downloading https://homebrew.bintray.com/bottles/clingo-5.4.0_2.catalina.bottle.tar.gz
Already downloaded: /Users/me/Library/Caches/Homebrew/downloads/84f6fc12ad562a91cee3984c937f708316a1e4e3b1634f7f95fe6f71f817f238--clingo-5.4.0_2.catalina.bottle.tar.gz
Error: clingo 4.5.4 is already installed
To upgrade to 5.4.0_2, run `brew upgrade clingo`.
Command that failed
07:34:36$ brew upgrade clingo
==> Upgrading 1 outdated package:
clingo 5.4.0_1 -> 5.4.0_2
==> Upgrading clingo 5.4.0_1 -> 5.4.0_2
==> Downloading https://homebrew.bintray.com/bottles/clingo-5.4.0_2.catalina.bottle.tar.gz
Already downloaded: /Users/me/Library/Caches/Homebrew/downloads/84f6fc12ad562a91cee3984c937f708316a1e4e3b1634f7f95fe6f71f817f238--clingo-5.4.0_2.catalina.bottle.tar.gz
Error: clingo 4.5.4 is already installed
To upgrade to 5.4.0_2, run `brew upgrade clingo`.
Output of command with --verbose --debug
07:35:54$ brew upgrade clingo --force --verbose --debug
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/clingo.rb
==> Upgrading 1 outdated package:
clingo 5.4.0_1 -> 5.4.0_2
==> Upgrading clingo 5.4.0_1 -> 5.4.0_2
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/cmake.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/sphinx-doc.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/python@3.9.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pkg-config.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/gdbm.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/openssl@1.1.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/readline.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/sqlite.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/xz.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/doxygen.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/bison.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/lua.rb
==> Downloading https://homebrew.bintray.com/bottles/clingo-5.4.0_2.catalina.bottle.tar.gz
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.6.0-39-g3821d37\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.7\)\ curl/7.64.1 --header Accept-Language:\ en --retry 3 --location --silent --head --request GET https://homebrew.bintray.com/bottles/clingo-5.4.0_2.catalina.bottle.tar.gz
Already downloaded: /Users/me/Library/Caches/Homebrew/downloads/84f6fc12ad562a91cee3984c937f708316a1e4e3b1634f7f95fe6f71f817f238--clingo-5.4.0_2.catalina.bottle.tar.gz
==> Verifying checksum for '84f6fc12ad562a91cee3984c937f708316a1e4e3b1634f7f95fe6f71f817f238--clingo-5.4.0_2.catalina.bottle.tar.gz'.
rm /usr/local/bin/clasp
rm /usr/local/bin/lpconvert
Error: clingo 4.5.4 is already installed
To upgrade to 5.4.0_2, run `brew upgrade clingo`.
ln -s ../Cellar/clasp/3.3.2/bin/clasp clasp
ln -s ../Cellar/clasp/3.3.2/bin/lpconvert lpconvert
Output of brew doctor --verbose
08:37:11$ brew doctor --verbose
==> Cask Environment Variables:
BUNDLE_PATH
CHRUBY_VERSION
GEM_HOME
GEM_PATH
HOMEBREW_CASK_OPTS
LC_ALL
PATH
RBENV_VERSION
RUBYLIB
RUBYOPT
RUBYPATH
SHELL
==> $LOAD_PATHS
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/sorbet-runtime-stub-0.2.0/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/ruby-macho-2.5.0/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/rubocop-sorbet-0.5.1/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/rubocop-rspec-2.0.0/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/rubocop-rails-2.8.1/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/rubocop-performance-1.9.1/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/ruby-progressbar-1.10.1/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/rack-2.2.3/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/plist-3.5.0/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/patchelf-1.3.0/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/mechanize-2.7.6/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/elftools-1.1.3/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/simplecov_json_formatter-0.1.2/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/bindata-2.4.8/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/activesupport-6.0.3.4/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/zeitwerk-2.4.1/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/tzinfo-1.2.8/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/thread_safe-0.3.6/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/i18n-1.8.5/lib
/usr/local/Homebrew/Library/Homebrew/vendor/bundle/bundler/../ruby/2.6.0/gems/concurrent-ruby-1.1.7/lib/concurrent-ruby
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/lib/ruby/2.6.0
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/lib/ruby/2.6.0/x86_64-darwin14
/usr/local/Homebrew/Library/Homebrew
==> Homebrew Version
2.6.0-42-g68889b9
==> macOS
10.15.7
==> SIP
Enabled
==> Java
1.8.0_275, 1.8.0_66
==> Homebrew Cask Staging Location
/usr/local/Caskroom
==> Homebrew Cask Taps:
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask (3732 casks)
/usr/local/Homebrew/Library/Taps/railwaycat/homebrew-emacsmacport (2 casks)
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask-versions (167 casks)
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask-drivers (166 casks)
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed with `brew diy`.
You should find replacements for the following formulae:
  python@2
  android-platform-tools
  camlp4
  wine
  ghc@8.2
  smlnj

Warning: "config" scripts exist outside your system or Homebrew directories.
`./configure` scripts often look for *-config scripts to determine if
software packages are installed, and which additional flags to use when
compiling and linking.

Having additional scripts in your path can confuse software installed via
Homebrew if the config script overrides a system or Homebrew-provided
script of the same name. We found the following "config" scripts:
  /Applications/azprolog.app/Contents/az_home/bin/azprolog-config
Error: unknown or unsupported macOS version: :mountain_lion
Output of brew tap
08:37:19$ brew tap
github/gh
homebrew/cask
homebrew/cask-drivers
homebrew/cask-versions
homebrew/core
railwaycat/emacsmacport

**Comment by Rylan12** [comment]
Can you paste the output of `brew info clingo`? I have a suspicion that this is related to a `clingo`/`gringo`/`clasp` issue that's popped up a few times. Make sure that you don't have `gringo` or `clasp` installed as well (`brew uninstall clasp gringo`) as they all point to the same formula. See https://github.com/Homebrew/discussions/discussions/117 and https://github.com/Homebrew/homebrew-core/issues/61522

(I haven't really read through the error message that closely so it's possible that this is something else)

**Comment by jasonhemann** [comment, marked as answer]
Thanks, that's something helpful and productive in the right direction.

```
16:12:04$ brew info clingo 
clingo: stable 5.4.0 (bottled)
ASP system to ground and solve logic programs
https://potassco.org/
/usr/local/Cellar/clasp/3.3.2 (7 files, 1.6MB) *
  Poured from bottle on 2017-10-18 at 22:13:47
/usr/local/Cellar/gringo/4.5.4 (7 files, 5.7MB) *
  Poured from bottle on 2017-10-18 at 22:13:43
/usr/local/Cellar/clingo/5.4.0_1 (22 files, 6.4MB)
  Poured from bottle on 2020-06-16 at 12:11:50
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/clingo.rb
License: MIT
==> Dependencies
Build: cmake ✔, doxygen ✔
Required: lua ✔, python@3.9 ✔
==> Analytics
install: 970 (30 days), 2,593 (90 days), 9,041 (365 days)
install-on-request: 591 (30 days), 1,473 (90 days), 4,012 (365 days)
build-error: 0 (30 days)
```
I did go and try that uninstall, but I think that's not doing what you'd hoped it would do. 

```
22:05:00$ brew uninstall clasp gringo 
Error: Refusing to uninstall /usr/local/Cellar/clingo/5.4.0_1
because it is required by aspcud, which is currently installed.
You can override this and force removal with:
  brew uninstall --ignore-dependencies clasp gringo
```

**Reply by Rylan12** [reply]
You could `brew uninstall --ignore-dependencies clasp gringo clingo` and then `brew install clingo`? Not familiar with `aspcud` but I doubt that will break anything. Alternatively, you could temporary `brew uninstall aspcud`

**Reply by jasonhemann** [reply]
I hadn't, but just did. 

```
08:53:47$ brew uninstall --ignore-dependencies clasp gringo
Uninstalling /usr/local/Cellar/clasp/3.3.2... (7 files, 1.6MB)
Uninstalling /usr/local/Cellar/clingo/5.4.0_1... (22 files, 6.4MB)
22:52:33$ brew uninstall --ignore-dependencies clingo
Error: No available formula or cask with the name "clingo".
22:52:43$ brew uninstall aspcud
Uninstalling /usr/local/Cellar/aspcud/1.9.4_1... (11 files, 445.5KB)
```

This did not appreciably help, AFAIK.

```
22:53:08$ brew install clingo
Updating Homebrew...
==> Auto-updated Homebrew!
Updated Homebrew from 68889b972 to edb749663.
Updated 4 taps (homebrew/cask-versions, homebrew/core, homebrew/cask and homebrew/cask-drivers).
...
==> Updated Formulae
Updated 26 formulae.
...
==> Deleted Casks
...

==> Downloading https://homebrew.bintray.com/bottles/clingo-5.4.0_2.catalina.bottle.tar.gz
Already downloaded: /Users/me/Library/Caches/Homebrew/downloads/84f6fc12ad562a91cee3984c937f708316a1e4e3b1634f7f95fe6f71f817f238--clingo-5.4.0_2.catalina.bottle.tar.gz
Error: clingo 4.5.4 is already installed
To upgrade to 5.4.0_2, run `brew upgrade clingo`.
```

**Reply by Rylan12** [reply]
Did you try `brew upgrade clingo` this time? If that doesn't work, how about `brew reinstall clingo`. If not, can I see `brew info clingo` again?

**Reply by himansri** [reply]
Jason,

I had the same issue , `brew info clingo` was not helping in troubleshooting.  
It turns out  (atleast in my case), above error occurs because clingo installer does not handle gracefully the preexistence of a component of clingo , and if it finds something , then it assumes clingo is already installed . (in my case, I had clasp folder in Cellar ... so doing 
`brew uninstall clasp`
followed by
`brew upgrade --verbose --debug clingo`
resolved the issue and the upgrade to clingo 5.4.0_3 went smoothly.

**Reply by jasonhemann** [reply]
That's precisely what I needed! Thank you!
